File: plugins/kraken_method_channel/ios/Classes/KrakenMethodChannelPlugin.m
#import "KrakenMethodChannelPlugin.h"
#import "KrakenMethodChannel.h"

@implementation KrakenMethodChannelPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"kraken_method_channel"
            binaryMessenger:[registrar messenger]];
  KrakenMethodChannelPlugin* instance = [[KrakenMethodChannelPlugin alloc] init];
  [[KrakenMethodChannel sharedMethodChannel] onAttach:instance channel:channel];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  [[KrakenMethodChannel sharedMethodChannel] handleMessageCall:call result:result];
}
@end

File: plugins/kraken_method_channel/macos/Classes/KrakenMethodChannelPlugin.swift
import Cocoa
import FlutterMacOS

public class KrakenMethodChannelPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "kraken_method_channel", binaryMessenger: registrar.messenger)
    let instance = KrakenMethodChannelPlugin()
    KrakenMethodChannel.shared.onAttach(KrakenMethodChannelPlugin: instance, channel: channel)
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    KrakenMethodChannel.shared.handleMessageCall(call, result: result)
  }
}

func defaultHandler (_ call: FlutterMethodCall, result: @escaping FlutterResult) {}

public class KrakenMethodChannel: NSObject {
  private static let instance:KrakenMethodChannel = KrakenMethodChannel()
  private var krakenMethodChannelPlugin:KrakenMethodChannelPlugin?
  private var channel:FlutterMethodChannel?
  private var handler = defaultHandler;
  
  private override init() {
  }
  
  public static var shared:KrakenMethodChannel {
    return self.instance
  }
  
  public func invokeMethod(method: String, arguments: Any?) {
    if (self.krakenMethodChannelPlugin != nil) {
      DispatchQueue.main.async {
        self.channel?.invokeMethod(method, arguments: arguments)
      }
    }
  }
  
  public func setMessageHandler(_ handler: @escaping (_ call: FlutterMethodCall, _ result: @escaping FlutterResult) -> Void) {
    self.handler = handler
  }
  
  public func handleMessageCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    self.handler(call, result)
  }
  
  public func onAttach(KrakenMethodChannelPlugin:KrakenMethodChannelPlugin, channel:FlutterMethodChannel) {
    self.krakenMethodChannelPlugin = KrakenMethodChannelPlugin
    self.channel = channel
  }
}
